### PR TITLE
CASSANDRA-18714 5.0 squashed SAI indexes for CQLSSTableWriter

### PR DESCRIPTION
**Pull request by smiklosovic**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java [hunk 1]
```diff
@@ -618,6 +653,29 @@ public CQLSSTableWriter build()
...
+                                                                                          directories,
+                                                                                          false,
+                                                                                          false,
+                                                                                          true);
+                        ks.initCfCustom(cfs);
+
+                        // this is the empty directory / leftover from times we initialized ColumnFamilyStore
+                        // it will automatically create directories for keyspace and table on disk after initialization
+                        // we set that directory to the destination of generated SSTables so we just remove empty directories here
+                        new File(directory, keyspaceName).deleteRecursive();
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java:731]: I put this back.

File: src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java [hunk 2]
```diff
@@ -677,6 +732,22 @@ public CQLSSTableWriter build()
...
+                    currentTables.put(Pair.create(tableMetadata.keyspace, tableMetadata.name), Pair.create(tableMetadata, new AtomicInteger(1)));
+                }
+                else
+                {
+                    Pair<TableMetadata, AtomicInteger> entry = currentTables.get(Pair.create(tableMetadata.keyspace, tableMetadata.name));
+                    if (entry != null)
+                    {
+                        if (!entry.left.equals(tableMetadata))
+                        {
+                            throw new IllegalStateException();
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java:744]: add exception message

File: src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java [hunk 3]
```diff
@@ -614,6 +655,19 @@ public Builder withBuildIndexes(boolean buildIndexes)
...
         }
 
+        /**
+         * Whether the underlying schema which was created for the purpose of the generation should be
+         * removed when the writing is finished. Defaults to true and should not be changed under normal circumstances.
+         *
+         * @param removeSchemaAfterDone true if underlying schema will be removed, false otherwise.
+         * @return this builder
+         */
+        public Builder removeSchemaAfterDone(boolean removeSchemaAfterDone)
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java:665]: we could have this package protected, it is used from tests only anyway

Comment (JeetKunDoug) [reply src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java:665]: Remember that the builder is a public interface for consumers outside of the C* code tree... this should stay public even if it's only used by tests in C*'s repository.

File: test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java [hunk 4]
```diff
@@ -63,38 +77,29 @@
...
-    private String table;
-    private String qualifiedTable;
-    private File dataDir;
+    private static final int NUMBER_WRITES_IN_RUNNABLE = 10;
 
-    static
-    {
-        DatabaseDescriptor.daemonInitialization();
-    }
+    private Supplier<CQLSSTableWriter.Builder> defaultBuilderSupplier = () -> CQLSSTableWriter.builder().removeSchemaAfterDone(false);
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java:86]:
```suggestion
    private final Supplier<CQLSSTableWriter.Builder> defaultBuilderSupplier = () -> CQLSSTableWriter.builder().removeSchemaAfterDone(false);
```

File: test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java [hunk 5]
```diff
@@ -63,38 +77,29 @@
...
-        Keyspace.setInitialized();
-        StorageService.instance.initServer();
-    }
+    private String keyspace;
+    protected String table;
+    private String qualifiedTable;
+    protected File dataDir;
 
     @Before
     public void perTestSetup() throws IOException
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java:97]:
```suggestion
    public void perTestSetup()
```

File: src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java [hunk 6]
```diff
@@ -128,14 +141,19 @@ public class CQLSSTableWriter implements Closeable
     private final ModificationStatement modificationStatement;
     private final List<ColumnSpecification> boundNames;
     private final List<TypeCodec> typeCodecs;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java:143]:
```suggestion
    private final List<TypeCodec<?>> typeCodecs;
```

**Review by maedhroz — APPROVED** [review]
Overall, LGTM

The only thing I'm not entirely sure of is whether we want the complexity of tracking `currentTables` outside the already existing static `Schema` state. Given the usage pattern of sequentially creating writers within the same JVM, we might consider an alternative where we just, rather than specifying `removeSchemaAfterDone`, tell the writer builder whether we want to overwrite existing/conflicting schema.

**Pull request closed without merging**